pkg/apis/networking/v1alpha1/clusteringress_types.go

```diff
@@ -0,0 +1,376 @@
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// TLS configuration. Currently the ClusterIngress only supports a single TLS
+	// port, 443. If multiple members of this list specify different hosts, they
+	// will be multiplexed on the same port according to the hostname specified
+	// through the SNI TLS extension, if the ingress controller fulfilling the
+	// ingress supports SNI.
+	// +optional
+	TLS []ClusterIngressTLS `json:"tls,omitempty"`
+
+	// A list of host rules used to configure the ClusterIngress.
+	// +optional
+	Rules []ClusterIngressRule `json:"rules,omitempty"`
+}
+
+// ClusterIngressTLS describes the transport layer security associated with an ClusterIngress.
+type ClusterIngressTLS struct {
+	// Hosts are a list of hosts included in the TLS certificate. The values in
+	// this list must match the name/s used in the tlsSecret. Defaults to the
+	// wildcard host setting for the loadbalancer controller fulfilling this
+	// ClusterIngress, if left unspecified.
+	// +optional
+	Hosts []string `json:"hosts,omitempty"`
+	// SecretName is the name of the secret used to terminate SSL traffic.
+	// Field is left optional to allow SSL routing based on SNI hostname alone.
+	// +optional
+	SecretName string `json:"secretName,omitempty"`
+
+	// SecretNamespace is the namespace of the secret used to terminate SSL traffic.
+	// Field is left optional to allow SSL routing based on SNI hostname alone.
+	// +optional
+	SecretNamespace string `json:"secretNamespace,omitempty"`
+
+	// ServerCertificate identifies the certificate filename in the secret.
+	// Defaults to `tls.cert`.
+	// +optional
+	ServerCertificate string `json:"serverCertificate,omitempty"`
+
+	// PrivateKey identifies the private key filename in the secret.
+	// Defaults to `tls.key`.
+	// +optional
+	PrivateKey string `json:"privateKey,omitempty"`
+}
+
+// ConditionType represents a ClusterIngress condition value
+const (
+	// ClusterIngressConditionReady is set when the clusterIngress networking setting is
+	// configured and it has a load balancer address.
+	ClusterIngressConditionReady = duckv1alpha1.ConditionReady
+
+	// ClusterIngressConditionNetworkConfigured is set when the ClusterIngress's underlying
+	// network programming has been configured.  This doesn't include conditions of the
+	// backends, so even if this should remain true when network is configured and backends
+	// are not ready.
+	ClusterIngressConditionNetworkConfigured duckv1alpha1.ConditionType = "NetworkConfigured"
+
+	// ClusterIngressConditionLoadBalancerReady is set when the ClusterIngress has
+	// a ready LoadBalancer.
+	ClusterIngressConditionLoadBalancerReady duckv1alpha1.ConditionType = "LoadBalancerReady"
+)
+
+var clusterIngressCondSet = duckv1alpha1.NewLivingConditionSet(
+	ClusterIngressConditionNetworkConfigured,
+	ClusterIngressConditionLoadBalancerReady)
+
+// ClusterIngressStatus describe the current state of the ClusterIngress.
+type ClusterIngressStatus struct {
```

Comment:
The ordering of types in this file is pretty haphazard. You have:

Top-level type
List type
Spec
Spec sub-message
Conditions
Status
Status sub-message
Spec sub-message